src/main/java/bio/terra/service/auth/iam/IamService.java

```diff
@@ -148,6 +149,39 @@ public Map<UUID, Set<IamRole>> listAuthorizedResources(
     return callProvider(() -> iamProvider.listAuthorizedResources(userReq, iamResourceType));
   }
 
+  /**
+   * Note: calling this method will trigger a call to SAM API. No cache backs it. If calling,
+   * consider whether new usage patterns necessitate refactoring to add a cache.
+   *
+   * @param userReq authenticated user
+   * @param iamResourceType resource type; e.g. dataset
+   * @param resourceId UUID identifying a resource
+   * @param actions required
+   * @throws IamForbiddenException if the user is missing any `actions` on the resource
+   */
+  public void verifyAuthorizations(
+      AuthenticatedUserRequest userReq,
+      IamResourceType iamResourceType,
+      String resourceId,
+      Collection<IamAction> actions)
+      throws IamForbiddenException {
+    String userEmail = userReq.getEmail();
+    List<String> availableActions =
+        callProvider(() -> iamProvider.listActions(userReq, iamResourceType, resourceId));
+
+    List<String> unavailableActions =
+        actions.stream()
+            .map(IamAction::toString)
+            .filter(action -> !availableActions.contains(action))
+            .toList();
+
+    if (!unavailableActions.isEmpty()) {
+      throw new IamForbiddenException(
```

Comment:
Example returned value (my Broad user did not have access to the dataset it was trying to patch):

```
{
  "message": "User 'okotsopo@broadinstitute.org' is missing required actions (returned in details)",
  "errorDetail": [
    "update_passport_identifier",
    "manage_schema"
  ]
}
```

Reply:
That's a clever use of the error serialization!